File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map <Integer, Integer> freq = new HashMap <Integer, Integer> ();
        List<List<Integer>> buckets = new ArrayList<List<Integer>> ();
        int [] sol = new int [k];
        int soli = 0;
        
        //created freq map
        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num,0)+1);
        }
        
        //created our buckets
        for (int i = 0; i <= nums.length; i++) {
            buckets.add(new ArrayList<Integer> ());
        }
        
        //populate buckets
        for (int elem : freq.keySet()) {
            int frequency = freq.get(elem);
            buckets.get(frequency).add(elem);
        }
        
        for (int i = buckets.size()-1; i >= 0; i--) {
            for (int elem : buckets.get(i)) {
                sol[soli] = elem;
                soli++;
                if (soli == k) break;
            } //for
            if (soli == k) break;
        } //for
        
        return sol;
    } //topK
} //Sol

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        #create our buckets
        buckets = [[] for i in range(len(nums) + 1)]
        #create freq map
        freq = {}
        #make sol list
        sol = []
        
        #calculate frequencies
        for num in nums:
            if num in freq:
                freq[num] = freq[num] + 1
            else:
                freq[num] = 1
                
        #put freqs into buckets
        for num in freq:
            buckets[freq[num]].append(num)
        
        #iterate backwards till k elements found
        for i in range(len(buckets) - 1, -1, -1):
            for num in buckets[i]:
                sol.append(num)
            if len(sol) == k:
                return sol
        